00-programming-fundamentals/intro-to-oauth.md

```diff
@@ -0,0 +1,46 @@
+
+# Introduction to OAuth
+
+## Learning Goal
+You should be able to:
+*  Explain the OAuth workflow and how an OAuth service can provide authentication and information about a user.
+
+## Introduction
+
+OAuth is an authorization framework, meaning it allows a user to grant access to their user account in one service (the OAuth provider) to another (the client).  At Ada we will primarily use OAuth to use another service (Github) to provide secure authentication to our sites.  
+
```

Comment:
This sentence is a little confusing to me, maybe it could be reworded. It might be good to provide definitions of _authentication_ and _authorization_ here, and to describe how OAuth complements what we've done so far. Perhaps also provide an example of a live webapp that allows login via OAuth. Something like:

```markdown
Today we will introduce a technology called *OAuth*, which we will use to allow users to log into our Rails app. OAuth is short for Open Authentication, and it is an open standard that describes how users can use one service to log into another. If you've ever logged into a website using Facebook, you've used OAuth.

### Authentication and Authorization

Security is a deep and nuanced field, but for our purposes there are two important pieces to get right:
- **Authentication:** Who are you, and how can I trust you?
- **Authorization:** Are you allowed to do that action?

Our _authorization_ story so far is a strong one. Once we know who a user is, we can use techniques like controller filters and validations to restrict actions they can take.

However, our _authentication_ story is quite weak. We trust the user to log in as whomever they want, without any restrictions.

One way to improve our authentication story would be to require the user to provide a secret password whenever they log in, so that only the person who created an account can log into it. However, handling passwords securely is a tricky business, and mistakes can have dire consequences for our users.

Instead we will rely on a trusted third party (GitHub) to authenticate users for us. Essentially, when a user tries to log in, we ask them to log into their GitHub account, and have GitHub send us a special token verifying the user is who they say they are. GitHub handles all the messy details of password management, not us.
```